### PR TITLE
Rewrite to use LO API SDK

### PR DESCRIPTION
**Pull request by jdys-at-orange**

### CODE REVIEW

File: src/main/java/com/orange/lo/sample/kerlink2lo/IotDeviceManagement.java [hunk 1]
```diff
@@ -24,33 +30,24 @@
...
-    private Map<String, KerlinkApi> kerlinkApiMap;
-    private LoDeviceProvider loDeviceProvider;
-    private ExternalConnectorService externalConnectorService;
-    private LoDeviceCache deviceCache;
+    private final KerlinkPropertiesList kerlinkPropertiesList;
+    private final Map<String, KerlinkApi> kerlinkApiMap;
+    private final LoDeviceProvider loDeviceProvider;
+    private final ExternalConnectorService externalConnectorService;
+    private final LoDeviceCache deviceCache;
+    private static final int synchronizationThreadPoolSize = 10;
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/IotDeviceManagement.java:44]: Constant names should comply with a naming convention.

File: src/main/java/com/orange/lo/sample/kerlink2lo/IotDeviceManagement.java [hunk 2]
```diff
@@ -7,13 +7,19 @@
...
 package com.orange.lo.sample.kerlink2lo;
 
-import com.orange.lo.sample.kerlink2lo.kerlink.KerlinkPropertiesList;
 import com.orange.lo.sample.kerlink2lo.kerlink.KerlinkApi;
-import com.orange.lo.sample.kerlink2lo.kerlink.model.EndDeviceDto;
+import com.orange.lo.sample.kerlink2lo.kerlink.KerlinkPropertiesList;
 import com.orange.lo.sample.kerlink2lo.lo.ExternalConnectorService;
 import com.orange.lo.sample.kerlink2lo.lo.LoDeviceCache;
 import com.orange.lo.sample.kerlink2lo.lo.LoDeviceProvider;
 import com.orange.lo.sample.kerlink2lo.lo.LoProperties;
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/IotDeviceManagement.java:15]: Unnecessary imports should be removed.

File: src/main/java/com/orange/lo/sample/kerlink2lo/kerlink/KerlinkApi.java [hunk 3]
```diff
@@ -1,48 +1,39 @@
...
 import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.core.ParameterizedTypeReference;
-import org.springframework.http.HttpEntity;
-import org.springframework.http.HttpHeaders;
-import org.springframework.http.HttpMethod;
-import org.springframework.http.HttpStatus;
-import org.springframework.http.ResponseEntity;
+import org.springframework.http.*;
 import org.springframework.scheduling.annotation.EnableScheduling;
 import org.springframework.scheduling.annotation.Scheduled;
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/kerlink/KerlinkApi.java:17]: Unnecessary imports should be removed.

File: src/main/java/com/orange/lo/sample/kerlink2lo/kerlink/KerlinkApi.java [hunk 4]
```diff
@@ -1,48 +1,39 @@
...
 import org.springframework.web.client.RestTemplate;
 
+import java.lang.invoke.MethodHandles;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
 @EnableScheduling
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/kerlink/KerlinkApi.java:27]: Why are you adding @EnableScheduling when you don't run anything?

File: src/main/java/com/orange/lo/sample/kerlink2lo/kerlink/KerlinkConfig.java [hunk 5]
```diff
@@ -1,24 +1,26 @@
...
-import java.util.HashMap;
-import java.util.Map;
-
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.web.client.RestTemplate;
 import org.springframework.web.util.DefaultUriBuilderFactory;
 
+import java.util.ArrayList;
+import java.util.List;
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/kerlink/KerlinkConfig.java:16]: Unnecessary imports should be removed.

File: src/main/java/com/orange/lo/sample/kerlink2lo/kerlink/KerlinkPropertiesList.java [hunk 6]
```diff
@@ -3,19 +3,20 @@
 import java.util.List;
 
 import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.boot.context.properties.ConstructorBinding;
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/kerlink/KerlinkPropertiesList.java:6]: Unnecessary imports should be removed.

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoApiClientWrapper.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package com.orange.lo.sample.kerlink2lo.lo;
+
+import com.orange.lo.sdk.LOApiClient;
+import org.springframework.context.annotation.DependsOn;
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoApiClientWrapper.java:4]: Unnecessary imports should be removed.

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoApiClientWrapper.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+public class LoApiClientWrapper {
+
+    private final LOApiClient loApiClient;
+
+    public LoApiClientWrapper(LOApiClient loApiClient) {
+        this.loApiClient = loApiClient;
+    }
+
+    @PostConstruct
+    public void Connect() {
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoApiClientWrapper.java:19]: Method names should comply with a naming convention.

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoConfig.java [hunk 9]
```diff
@@ -1,38 +1,61 @@
...
+                .keepAliveIntervalSeconds(loProperties.getKeepAliveIntervalSeconds())
+                .mqttPersistenceDataDir(loProperties.getMqttPersistenceDir())
+                .dataManagementExtConnectorCommandCallback(dataManagementExtConnectorCommandCallback)
+                .build();
+    }
 
-    public LoConfig(LoProperties loProperties) {
-        this.loProperties = loProperties;
+    @Bean
+    public LOApiClient LOApiClient(LOApiClientParameters parameters) {
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoConfig.java:43]: Method names should comply with a naming convention.

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoConfig.java [hunk 10]
```diff
@@ -1,38 +1,61 @@
...
+ */
 
 package com.orange.lo.sample.kerlink2lo.lo;
 
+import com.orange.lo.sample.kerlink2lo.kerlink.KerlinkApi;
+import com.orange.lo.sdk.LOApiClient;
+import com.orange.lo.sdk.LOApiClientParameters;
+import com.orange.lo.sdk.externalconnector.DataManagementExtConnector;
+import com.orange.lo.sdk.externalconnector.DataManagementExtConnectorCommandCallback;
+import com.orange.lo.sdk.fifomqtt.DataManagementFifoCallback;
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoConfig.java:15]: Unnecessary imports should be removed.

File: pom.xml [hunk 11]
```diff
@@ -69,9 +86,96 @@
 	<build>
 		<plugins>
 			<plugin>
-				<groupId>org.springframework.boot</groupId>
-				<artifactId>spring-boot-maven-plugin</artifactId>
+				<groupId>org.apache.maven.plugins</groupId>
```
Comment (grzegorzswistak) [inline pom.xml:89]: Adding the maven release plugin is not part of this sentence. However, if you want to do it, you should add <scm> with the correct values and a workflow with manual release.

File: src/main/java/com/orange/lo/sample/kerlink2lo/kerlink/KerlinkLoginCaller.java [hunk 12]
```diff
@@ -0,0 +1,23 @@
+package com.orange.lo.sample.kerlink2lo.kerlink;
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/kerlink/KerlinkLoginCaller.java:1]: This class should contain the license header.

File: src/main/java/com/orange/lo/sample/kerlink2lo/kerlink/KerlinkPropertiesList.java [hunk 13]
```diff
@@ -3,19 +3,20 @@
 import java.util.List;
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/kerlink/KerlinkPropertiesList.java:3]: This class should contain the license header.

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoApiClientWrapper.java [hunk 14]
```diff
@@ -0,0 +1,22 @@
+package com.orange.lo.sample.kerlink2lo.lo;
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoApiClientWrapper.java:1]: This class should contain the license header.

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoProperties.java [hunk 15]
```diff
@@ -1,125 +1,74 @@
-/**
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoProperties.java:1]: This class should contain the license header.

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/MessageListener.java [hunk 16]
```diff
@@ -0,0 +1,69 @@
+package com.orange.lo.sample.kerlink2lo.lo;
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/MessageListener.java:1]: This class should contain the license header.

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/PayloadDecoder.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
+package com.orange.lo.sample.kerlink2lo.lo;
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/PayloadDecoder.java:1]: This class should contain the license header.

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/PayloadDecoderFactory.java [hunk 18]
```diff
@@ -0,0 +1,25 @@
+package com.orange.lo.sample.kerlink2lo.lo;
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/PayloadDecoderFactory.java:1]: This class should contain the license header.

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/PayloadDecoderFactory.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.stereotype.Component;
+
+@Component
+public class PayloadDecoderFactory {
+    @Bean
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/PayloadDecoderFactory.java:8]: Why are you using the @Bean annotation here?

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/PayloadDecoderFactory.java [hunk 20]
```diff
@@ -0,0 +1,25 @@
+package com.orange.lo.sample.kerlink2lo.lo;
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.stereotype.Component;
+
+@Component
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/PayloadDecoderFactory.java:6]: Why are you adding a class to the context that you never inject?

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoApiExternalConnectorService.java [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+
+        if (connectorDecoderName.isPresent()) {
+            dataMessage.setMetadata(new Metadata(connectorDecoderName.get()));
+        }
+        loApiDataManagementExtConnector.sendMessage(deviceId, dataMessage);
+    }
+
+    @Override
+    public void sendCommandResponse(DataDownEventDto dataDownEventDto) {
+        externalConnectorCommandResponseWrapper.sendCommandResponse(dataDownEventDto);
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoApiExternalConnectorService.java:65]: Why do you use DataManagementExtConnector directly in other methods of this class and here you use a class that wraps it?

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoApiClientWrapper.java [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+
+    private final LOApiClient loApiClient;
+
+    public LoApiClientWrapper(LOApiClient loApiClient) {
+        this.loApiClient = loApiClient;
+    }
+
+    @PostConstruct
+    public void Connect() {
+        loApiClient.getDataManagementFifo().connectAndSubscribe();
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/LoApiClientWrapper.java:20]: How many mqtt topics do you subscribe to here? Connector only connects to mqtt topics via external connector.

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/ExternalConnectorService.java [hunk 23]
```diff
@@ -1,52 +1,14 @@
...
-* LICENSE file in the root directory of this source tree. 
-*/
-
 package com.orange.lo.sample.kerlink2lo.lo;
 
 import com.orange.lo.sample.kerlink2lo.kerlink.model.DataDownEventDto;
 import com.orange.lo.sample.kerlink2lo.kerlink.model.DataUpDto;
 
-import org.springframework.stereotype.Service;
+public interface ExternalConnectorService {
```
Comment (grzegorzswistak) [inline src/main/java/com/orange/lo/sample/kerlink2lo/lo/ExternalConnectorService.java:6]: What is the purpose of this interface?

Comment (jdys-at-orange) [reply src/main/java/com/orange/lo/sample/kerlink2lo/lo/ExternalConnectorService.java:13]: Was used in testing, though not really necessary with Mockito. Will remove.

File: src/main/java/com/orange/lo/sample/kerlink2lo/lo/PayloadDecoderFactory.java (hunk 19 above)
Comment (jdys-at-orange) [reply src/main/java/com/orange/lo/sample/kerlink2lo/lo/PayloadDecoderFactory.java:15]: I misunderstood how is decoder applied at first. This function is no longer needed.

**Review by grzegorzswistak — APPROVED** [review]

**Pull request merged**